ADAL/src/validation/ADAadAuthorityCache.m

```diff
@@ -0,0 +1,274 @@
+                                       protocolCode:nil \
+                                       errorDetails:_ERROR \
+                                      correlationId:context.correlationId]; \
+        if (error) { *error = adError; } \
+        return NO; \
+    }
+
+@implementation ADAadAuthorityCacheRecord
+
+@end
+
+@implementation ADAadAuthorityCache
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    
+    _recordMap = [NSMutableDictionary new];
+    pthread_rwlock_init(&_rwLock, NULL);
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    pthread_rwlock_destroy(&_rwLock);
+}
+
+- (BOOL)processMetadata:(NSArray<NSDictionary *> *)metadata
+              authority:(NSURL *)authority
+                context:(id<ADRequestContext>)context
+                  error:(ADAuthenticationError * __autoreleasing *)error
+{
+    if (metadata != nil)
+    {
+        CHECK_CLASS_TYPE(metadata, NSArray, @"JSON metadata from authority validation is not an array");
+    }
+    
+    [self getWriteLock];
+    BOOL ret = [self processImpl:metadata authority:authority context:context error:error];
+    pthread_rwlock_unlock(&_rwLock);
+    
+    return ret;
+}
+
+- (BOOL)processImpl:(NSArray<NSDictionary *> *)metadata
+          authority:(NSURL *)authority
+            context:(id<ADRequestContext>)context
+              error:(ADAuthenticationError * __autoreleasing *)error
+{
+    if (metadata.count == 0)
+    {
+        AD_LOG_INFO(@"No metadata returned from authority validation", context.correlationId, nil);
+    }
+    else
+    {
+        AD_LOG_INFO(@"Caching AAD Environements:", context.correlationId, nil);
+    }
+    
+    NSMutableArray<ADAadAuthorityCacheRecord *> *recordsToAdd = [NSMutableArray new];
+    
+    for (NSDictionary *environment in metadata)
+    {
+        CHECK_CLASS_TYPE(environment, NSDictionary, @"JSON metadata entry is not a dictionary");
+        
+        __auto_type *record = [ADAadAuthorityCacheRecord new];
+        record.validated = YES;
+        
+        NSString *networkHost = environment[@"preferred_network"];
+        CHECK_CLASS_TYPE(networkHost, NSString, @"\"preferred_network\" in JSON authority validation metadata must be a string");
+        record.networkHost = networkHost;
+        
+        NSString *cacheHost = environment[@"preferred_cache"];
+        CHECK_CLASS_TYPE(cacheHost, NSString, @"\"preferred_cache\" in JSON authority validation metadata must be a string");
+        record.cacheHost = cacheHost;
+        
+        NSArray *aliases = environment[@"aliases"];
+        CHECK_CLASS_TYPE(aliases, NSArray, @"\"alias\" in JSON authority validation metadata must be an array");
+        record.aliases = aliases;
+        
+        for (NSString *alias in aliases)
+        {
+            CHECK_CLASS_TYPE(alias, NSString, @"\"alias\" in JSON authority validation metadata must be an array of strings");
```

Comment:
This error should probably says "aliases" as that's what's in the spec as an array of strings.